### PR TITLE
+ [android] support prerender

### PR DESCRIPTION
**Pull request by Rowandjj**
introduction : https://issues.apache.org/jira/browse/WEEX-27?jql=project%20%3D%20WEEX%20AND%20resolution%20%3D%20Unresolved%20AND%20priority%20%3D%20Major%20ORDER%20BY%20key%20DESC

### CODE REVIEW

**Comment by Rowandjj** [comment]
@sospartan pls review it:)

File: android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java [hunk 1]
```diff
@@ -122,7 +123,8 @@
   /**
    * Render start time
    */
-  private long mRenderStartTime;
+  //this field may accessed outside
+  public long mRenderStartTime;
```
Comment (sospartan) [inline android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java]: no public field.

File: android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java [hunk 2]
```diff
@@ -625,6 +635,24 @@ public void reloadImages() {
...
+  public void setPreRenderMode(final boolean isPreRenderMode) {
+    WXSDKManager.getInstance().getWXRenderManager().postOnUiThread(new Runnable() {
+      @Override
+      public void run() {
+        WXSDKInstance.this.isPreRenderMode = isPreRenderMode;
+      }
+    },0);
+  }
+
+  public void replaceContext(@NonNull Context context) {
```
Comment (sospartan) [inline android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java]: 'setContext' is fair enough. Pls add more comments about why context is set-able

File: android/sdk/src/main/java/com/taobao/weex/dom/action/CreateFinishAction.java [hunk 3]
```diff
@@ -18,15 +18,36 @@
...
 /**
  * Created by sospartan on 02/03/2017.
  */
 
 final class CreateFinishAction extends AbstractLayoutFinishAction {
+
+  @Override
+  public void executeDom(DOMActionContext context) {
+    super.executeDom(context);
+    //add by chuyi
```
Comment (sospartan) [inline android/sdk/src/main/java/com/taobao/weex/dom/action/CreateFinishAction.java]: ?

File: android/sdk/src/main/java/com/taobao/weex/dom/action/ExecutableAction.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package com.taobao.weex.dom.action;
+
+import com.taobao.weex.dom.DOMAction;
+import com.taobao.weex.dom.DOMActionContext;
+import com.taobao.weex.dom.RenderAction;
+import com.taobao.weex.dom.RenderActionContext;
+
+class ExecutableAction implements DOMAction, RenderAction {
```
Comment (sospartan) [inline android/sdk/src/main/java/com/taobao/weex/dom/action/ExecutableAction.java]: This action post a runnable job to render thread, the name should be more clear. Better add 'render' to the class name.

File: android/sdk/src/main/java/com/taobao/weex/ui/module/WXTimerModule.java [hunk 5]
```diff
@@ -57,28 +59,32 @@ void setHandler(Handler handler) {
 
   @JSMethod(uiThread = false)
   public void setTimeout(@IntRange(from = 1) int funcId, @IntRange(from = 0) int delay) {
-    postMessage(WXJSBridgeMsgType.MODULE_TIMEOUT, funcId, delay, Integer.parseInt(mWXSDKInstance.getInstanceId()));
+    if(mWXSDKInstance != null) {
```
Comment (sospartan) [inline android/sdk/src/main/java/com/taobao/weex/ui/module/WXTimerModule.java]: This file is irrelevant to the PR.

File: android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java (hunk 1 above)
Comment (Rowandjj) [inline android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java]: ok i'll add setter method

File: android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java (hunk 2 above)
Comment (Rowandjj) [inline android/sdk/src/main/java/com/taobao/weex/WXSDKInstance.java]: ok

File: android/sdk/src/main/java/com/taobao/weex/dom/action/ExecutableAction.java (hunk 4 above)
Comment (Rowandjj) [inline android/sdk/src/main/java/com/taobao/weex/dom/action/ExecutableAction.java]: changed to ` ExecutableRenderAction`

File: android/sdk/src/main/java/com/taobao/weex/ui/module/WXTimerModule.java (hunk 5 above)
Comment (Rowandjj) [inline android/sdk/src/main/java/com/taobao/weex/ui/module/WXTimerModule.java]: nope... we want hold timers in preRenderMode

File: android/sdk/src/main/java/com/taobao/weex/dom/action/CreateFinishAction.java (hunk 3 above)
Comment (Rowandjj) [inline android/sdk/src/main/java/com/taobao/weex/dom/action/CreateFinishAction.java]: i want 'save cached instance' when layout finished. so i add `LayoutFinishListener`, when createFinish action invoked, the listener will fired too.

**Pull request merged**